Source/CombatExtended/CombatExtended/Comps/CompAmmosetSwitcher.cs

```diff
@@ -0,0 +1,262 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using RimWorld;
+using Verse;
+using UnityEngine;
+using CombatExtended.Utilities;
+
+namespace CombatExtended
+{
+    public class CompUnderBarrel : CompRangedGizmoGiver
+    {
+        public CompProperties_UnderBarrel Props => (CompProperties_UnderBarrel)this.props;
+
+        public CompEquippable compEq => this.parent.TryGetComp<CompEquippable>();
+
+        public CompAmmoUser compAmmo => this.parent.TryGetComp<CompAmmoUser>();
+
+        public CompFireModes compFireModes => this.parent.TryGetComp<CompFireModes>();
+
+        public CompProperties_FireModes compPropsFireModes => this.parent.def.comps.Find(x => x is CompProperties_FireModes) as CompProperties_FireModes;

```

Comment:
If any of these are going to be called a lot, and especially if the values do not (or rarely) change, they should be cached.  
If they can run multiple times per tick, but the value could change indirectly, you can at the least cache the result and invalidate it if the tick number increases.